Include stash, bisect in grid if tags are shown

gerhardol opened the pull request:
Fixes #8879

## Proposed changes

Revert behavior to be same as in 3.4

If tag labels are shown, also show bisect, notes, stash etc

Note that the labels were orange in 3.4, changed in #8424

## Screenshots <!-- Remove this section if PR does not change UI -->

### Before

(slightly older than master)
![image](https://user-images.githubusercontent.com/6248932/111052110-79c6eb00-8458-11eb-8eb9-dae9a26ccdc7.png)

### After

![image](https://user-images.githubusercontent.com/6248932/111052089-45ebc580-8458-11eb-849b-b07f086d8694.png)

## Test methodology <!-- How did you ensure quality? -->

Manual

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

RussKie commented:
Thank you.
Please let me tinker with it.

On Sun, 14 Mar. 2021, 10:04 am Gerhard Olsson, ***@***.***>
wrote:

> Fixes #8879 <https://github.com/gitextensions/gitextensions/issues/8879>
> Proposed changes
>
> Revert behavior to be same as in 3.4
>
> If tag labels are shown, also show bisect, notes, stash etc
>
> Note that the labels were orange in 3.4, changed in #8424
> <https://github.com/gitextensions/gitextensions/pull/8424>
>
> WIP as not tested at all
> Screenshots Before
>
> (slightly older than master)
> [image: image]
> <https://user-images.githubusercontent.com/6248932/111052110-79c6eb00-8458-11eb-8eb9-dae9a26ccdc7.png>
> After
>
> [image: image]
> <https://user-images.githubusercontent.com/6248932/111052089-45ebc580-8458-11eb-849b-b07f086d8694.png>
> Test methodology
>
> Manual
> ------------------------------
>
> ✒️ I contribute this code under The Developer Certificate of Origin
> <http://../blob/master/contributors.txt>.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/gitextensions/gitextensions/pull/8990
> Commit Summary
>
>    - Include stash, bisect in grid if tags are shown
>
> File Changes
>
>    - *M* GitCommands/Git/Commands/GitCommandHelpers.cs
>    <https://github.com/gitextensions/gitextensions/pull/8990/files#diff-c944249d8e62fae1b208e2294e9ed0fac1fd9b4142dd2186e03968d733060815>
>    (4)
>
> Patch Links:
>
>    - https://github.com/gitextensions/gitextensions/pull/8990.patch
>    - https://github.com/gitextensions/gitextensions/pull/8990.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/8990>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXVTMGYRVRRMSU5HY7LTDPVODANCNFSM4ZEJTURQ>
> .
>

RussKie commented on GitCommands/Git/Commands/GitCommandHelpers.cs:
```diff
@@ -109,9 +109,7 @@ public static ArgumentString GetRefsCmd(bool tags, bool branches, bool noLocks,
-                { branches, "refs/heads/", string.Empty },
-                { branches && tags, "refs/remotes/", string.Empty },
-                { tags, "refs/tags/", string.Empty },
+                { !tags, "refs/remotes/ refs/heads/", string.Empty },
```
This is incorrect, since this would never be true.

Since we can't filter at `` commands, we can fix it by delegating the filter to the bash:
```diff
                 format,
-                { branches, "refs/heads/", string.Empty },
-                { branches && tags, "refs/remotes/", string.Empty },
-                { tags, "refs/tags/", string.Empty },
+                { !branches, "| sed 'refs\\/heads\\//d", string.Empty },
+                { !(branches && tags), "| sed 'refs\\/remotes\\//d", string.Empty },
+                { !tags, "| sed 'refs\\/tags\\//d", string.Empty },
             };
```

RussKie replied:
Here's the fix for the tests:
```diff
diff --git a/UnitTests/GitCommands.Tests/Git/Commands/GitCommandHelpersTest.cs b/UnitTests/GitCommands.Tests/Git/Commands/GitCommandHelpersTest.cs
index ec2ab688c..62e35dee9 100644
--- a/UnitTests/GitCommands.Tests/Git/Commands/GitCommandHelpersTest.cs
+++ b/UnitTests/GitCommands.Tests/Git/Commands/GitCommandHelpersTest.cs
@@ -664,13 +664,14 @@ private static IEnumerable<TestCaseData> GetRefsCommandTestData
                         }
 
                         yield return new TestCaseData(/* tags */ true, /* branches */ true, /* noLocks */ false, sortBy, sortOrder,
-                            /* expected */ $@"for-each-ref{sortConditionRef}{sortCondition}{format} refs/heads/ refs/remotes/ refs/tags/");
+                            /* expected */ $@"for-each-ref{sortConditionRef}{sortCondition}{format}");
                         yield return new TestCaseData(/* tags */ true, /* branches */ false, /* noLocks */ false, sortBy, sortOrder,
-                            /* expected */ $@"for-each-ref{sortConditionRef}{sortCondition}{format} refs/tags/");
+                            /* expected */ $@"for-each-ref{sortConditionRef}{sortCondition}{format} | sed 'refs\/heads\//d | sed 'refs\/remotes\//d");
                         yield return new TestCaseData(/* tags */ false, /* branches */ true, /* noLocks */ false, sortBy, sortOrder,
-                            /* expected */ $@"for-each-ref{sortCondition} --format=""%(objectname) %(refname)"" refs/heads/");
+                            /* expected */ $@"for-each-ref{sortCondition} --format=""%(objectname) %(refname)"" | sed 'refs\/remotes\//d | sed 'refs\/tags\//d");
+
                         yield return new TestCaseData(/* tags */ false, /* branches */ true, /* noLocks */ true, sortBy, sortOrder,
-                            /* expected */ $@"--no-optional-locks for-each-ref{sortCondition} --format=""%(objectname) %(refname)"" refs/heads/");
+                            /* expected */ $@"--no-optional-locks for-each-ref{sortCondition} --format=""%(objectname) %(refname)"" | sed 'refs\/remotes\//d | sed 'refs\/tags\//d");
                     }
                 }
             }
```

gerhardol replied:
The argument need to be added twice, in some other way. But the matching should be done on a positional base, sed will match in the middle of the refs too

gerhardol replied:
I changed the solution to handle the RevGrid usage specially and explicitly, to really show that this is a special usage.
I looked at the usage for when both tags and branches were retrieved, they were all to get remotes - that should be a separate option too, but I do not want to change the IGitModule interface. Added some xml doc.

Note that stash, notes, bisect labels are always shown.

RussKie replied:
> The argument need to be added twice, in some other way. But the matching should be done on a positional base, sed will match in the middle of the refs too

To match the start of line use `^`

https://www.howtogeek.com/666395/how-to-use-the-sed-command-on-linux/

RussKie replied:
> The argument need to be added twice, in some other way. 

Can you please clarify this bit?

RussKie commented on GitCommands/Git/GitModule.cs:
```diff
@@ -2903,6 +2903,18 @@ public RemoteActionResult<IReadOnlyList<IGitRef>> GetRemoteServerRefs(string rem
+        /// <param name="branches">Include local branches, also remote branches if <see paramref="tags"/> is set.</param>
+        /// <param name="all">Include all refs (override <see paramref="tags"/>, <see paramref="branches"/>). This includes stash, bisect, stash.</param>
+        /// <returns>All Git refs.</returns>
+        public IReadOnlyList<IGitRef> GetAllRefs(bool tags = true, bool branches = true, bool all = false)
```
If we're going down this path then we ought to create an enum:

Local
Remote
Tag
All = Local | Remote | Tag

gerhardol replied:
All is more than that, it was the problem here...
The handling of remotes is quite gross too. 

Can we make the change like this now and add an enum later in master?

gerhardol replied:
Can we change the IGitModule interface here?

RussKie replied:
Yes

RussKie commented:
The more pocessing we can offload to the underlying cli - the better, we'll do less work, less allocations, which is a win for our users.

RussKie commented:
Sure

On Mon, 15 Mar. 2021, 10:21 am Gerhard Olsson, ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In GitCommands/Git/GitModule.cs
> <https://github.com/gitextensions/gitextensions/pull/8990#discussion_r593978465>
> :
>
> > @@ -2903,6 +2903,18 @@ public RemoteActionResult<IReadOnlyList<IGitRef>> GetRemoteServerRefs(string rem
>          }
>
>          public IReadOnlyList<IGitRef> GetRefs(bool tags = true, bool branches = true)
> +        {
> +            return GetAllRefs(tags, branches, all: false);
> +        }
> +
> +        /// <summary>
> +        /// Get the Git refs.
> +        /// </summary>
> +        /// <param name="tags">Include tags.</param>
> +        /// <param name="branches">Include local branches, also remote branches if <see paramref="tags"/> is set.</param>
> +        /// <param name="all">Include all refs (override <see paramref="tags"/>, <see paramref="branches"/>). This includes stash, bisect, stash.</param>
> +        /// <returns>All Git refs.</returns>
> +        public IReadOnlyList<IGitRef> GetAllRefs(bool tags = true, bool branches = true, bool all = false)
>
> All is more than that, it was the problem here...
> The handling of remotes is quite gross too.
>
> Can we make the change like this now and add an enum later in master?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/8990#discussion_r593978465>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXW7AR55G4LZNCPQMWTTDVAFZANCNFSM4ZEJTURQ>
> .
>

RussKie commented on GitCommands/Git/Commands/GitCommandHelpers.cs:
```diff
@@ -107,11 +108,11 @@ public static ArgumentString GetRefsCmd(bool tags, bool branches, bool noLocks,
-                { tags, "refs/tags/", string.Empty },
+                { (getRef & GetRefsEnum.All) == 0 && (getRef & GetRefsEnum.Branches) != 0, "refs/heads/", string.Empty },
+                { (getRef & GetRefsEnum.All) == 0 && (getRef & GetRefsEnum.Remotes) != 0, "refs/remotes/", string.Empty },
+                { (getRef & GetRefsEnum.All) == 0 && (getRef & GetRefsEnum.Tags) != 0, "refs/tags/", string.Empty },
```
We're duplicating this code, so it is worth considering extracting it in a local static method:
```cs
static bool HasFlag(GetRefsEnum requested, GetRefsEnum flag)
    => (requested & GetRefsEnum.All) == 0 && (requested & flag) != 0;
```

gerhardol replied:
changed

RussKie reviewed (approved):
👍

gerhardol commented:
Note that GetRefs are called six times during a refresh (probably four from the sidepanel).
The startup can be optimized...

Remotes were incorrectly handled in the latest commit, rebased and squashed.

RussKie commented on GitCommands/Git/Commands/GitCommandHelpers.cs:
```diff
@@ -79,10 +79,11 @@ public static ArgumentString AddSubmoduleCmd(string remotePath, string localPath
         {
             string format;
-            if (!tags)
+            bool hasTags = (getRef & GetRefsEnum.All) != 0 || (getRef & GetRefsEnum.Tags) != 0;
```
Did you miss this one?

gerhardol replied:
This is different - I can add a special if you prefer

RussKie commented:
No worries then

On Fri, 19 Mar. 2021, 8:07 am Gerhard Olsson, ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In GitCommands/Git/Commands/GitCommandHelpers.cs
> <https://github.com/gitextensions/gitextensions/pull/8990#discussion_r597242806>
> :
>
> >          {
>              string format;
> -            if (!tags)
> +            bool hasTags = (getRef & GetRefsEnum.All) != 0 || (getRef & GetRefsEnum.Tags) != 0;
>
> This is different - I can add a special if you prefer
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/8990#discussion_r597242806>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXURBMZJI4OTF363VBDTEJTRPANCNFSM4ZEJTURQ>
> .
>
